pkg/handler/collector/oci/oci.go

```diff
@@ -54,84 +54,94 @@ func NewOCICollector(ctx context.Context, repoTags map[string][]string, poll boo
-				}
-				// set interval to about 5 mins or more
-				time.Sleep(o.interval)
-			}
-		}
-	} else {
+	if docChannel == nil {
+		return fmt.Errorf("nil channel passed to %s", OCICollector)
+	}
+
+	for {
 		for repo, tags := range o.repoTags {
-			err := o.getTagsAndFetch(ctx, repo, tags, docChannel)
-			if err != nil {
-				return err
+			if len(tags) > 0 && o.poll {
+				return errors.New("image tag should not specified when using polling")
+			}
+			if err := o.getTagsAndFetch(ctx, repo, tags, docChannel); err != nil {
+				return fmt.Errorf("fetching artifacts for %s: %w", repo, err)
 			}
 		}
+		if !o.poll {
+			break
+		}
+		select {
+		case <-ctx.Done():
+			// This way the function will not block for a fixed
+			// interval and can be stopped immediately when the context is canceled.
+			return ctx.Err()
+		case <-time.After(o.interval):
+		}
 	}
 
 	return nil
 }
 
 func (o *ociCollector) getTagsAndFetch(ctx context.Context, repo string, tags []string, docChannel chan<- *processor.Document) error {
-	rcOpts := []regclient.Opt{}
-	rcOpts = append(rcOpts, regclient.WithDockerCreds())
-	rcOpts = append(rcOpts, regclient.WithDockerCerts())
+	if repo == "" {
+		return errors.New("repository name not specified")
+	}
 
-	if len(tags) > 0 {
-		for _, tag := range tags {
-			if tag == "" {
-				return errors.New("image tag not specified to fetch")
-			}
-			imageTag := fmt.Sprintf("%v:%v", repo, tag)
-			r, err := ref.New(imageTag)
-			if err != nil {
-				return err
-			}
+	if docChannel == nil {
+		return errors.New("invalid document channel")
+	}
+	excludedSuffixes := []string{"sbom", "att", "sig"}
 
-			rc := regclient.New(rcOpts...)
-			defer rc.Close(ctx, r)
+	rcOpts := []regclient.Opt{
+		regclient.WithDockerCreds(),
+		regclient.WithDockerCerts(),
+	}
 
-			err = o.fetchOCIArtifacts(ctx, repo, rc, r, docChannel)
-			if err != nil {
-				return err
-			}
-		}
-	} else {
-		r, err := ref.New(repo)
-		if err != nil {
-			return err
-		}
+	rc := regclient.New(rcOpts...)
 
-		rc := regclient.New(rcOpts...)
-		defer rc.Close(ctx, r)
+	repoRef, err := ref.New(repo)
+	defer rc.Close(ctx, repoRef)
 
-		tags, err := rc.TagList(ctx, r)
+	if err != nil {
+		return fmt.Errorf("parsing repository reference: %w", err)
+	}
+
+	var fetchTags []string
+	if len(tags) > 0 {
+		fetchTags = tags
+	} else {
+		allTags, err := rc.TagList(ctx, repoRef)
 		if err != nil {
 			return fmt.Errorf("reading tags for %s: %w", repo, err)
 		}
 
-		for _, tag := range tags.Tags {
-			if !strings.HasSuffix(tag, "sbom") && !strings.HasSuffix(tag, "att") && !strings.HasSuffix(tag, "sig") {
-				imageTag := fmt.Sprintf("%v:%v", repo, tag)
-				r, err := ref.New(imageTag)
-				if err != nil {
-					return err
-				}
-				err = o.fetchOCIArtifacts(ctx, repo, rc, r, docChannel)
-				if err != nil {
-					return err
+		// Filter out tags that are not images
+		for _, tag := range allTags.Tags {
+			shouldInclude := true
+			for _, suffix := range excludedSuffixes {
```

Comment:
Rather than hard coding a list, or having an array of suffixes, this would probably be best done with a prefix check, looking for `sha256-` followed by a 64 character hex string. There's no suffix with the OCI referrers fallback tag, and it possible someone might push data up with a different extension.

A more reliable check if something is an image is to look at the image manifest, checking the mediaType on the config and layers to make sure they match the media types used by images (there are several, compressed, uncompressed, foreign, plus docker and OCI variants of each). However that requires pulling the manifest for every tag (and nested manifest in a manifest list / index).

Reply:
Thanks, took the first approach which was much easier.